sdk/component/runner.go

```diff
@@ -0,0 +1,143 @@
+package component
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os/signal"
+	"syscall"
+
+	"golang.org/x/sync/errgroup"
+)
+
+type (
+	// runner is used to setup the run context.
+	runner struct {
+		config *RunnerConfig
+	}
+
+	// componentRunnerFunc is an alias for executing component run functions.
+	componentRunnerFunc func(context.Context, *RunnerConfig) error
+)
+
+// newRunner returns a new runner which is initialised looking at the default configuration
+// and can be customised applying options.
+// Returns an error if the configuration is invalid.
+func newRunner(opts ...RunnerOption) (*runner, error) {
+	cfg, err := newRunnerConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not create default runner configuration: %w", err)
+	}
+
+	r := &runner{config: cfg}
+
+	for _, opt := range opts {
+		if err := opt(r); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := r.config.isValid(); err != nil {
+		return nil, fmt.Errorf("invalid configuration: %w", err)
+	}
+
+	return r, nil
+}
+
+// run is the core of the component SDK.
+// run performs all the steps needed to run components reliably and predictably.
+// run takes care of:
+// - signal termination: gracefully close components when a termination signal is detected.
+// - context cancellation: gracefully close components when a cancellation is detected.
+// - panic handling.
+// - logging: initialisation and centralisation.
+// - running components: reliably and respecting cancellations.
+// - TODO: metrics.
+// - TODO: tracing.
+func run(
```

Comment:
This is the main brain and makes all the magic happen. The comment should be clear but let me know.